### PR TITLE
requireSpacesInForStatement: Requires spaces inbetween for statement.

### LINKED ISSUE #912: Rule: require spaces in between for statement
Requires spaces in between for statement.
 
  Type: `Boolean`
 
  Values: `true` to requires spaces in between for statement.
 
  #### Example
 
  ```js
  "requireSpacesInForStatement": true
  ```
 
  ##### Valid
 
  ```js
  for(var i = 0; i<l; i++) {
      x++;
  }
  ```
 
  ##### Invalid
 
  ```js
  for(var i = 0;i<l;i++) {
      x++;
  }
  ```
 
  ```js
  for(var i = 0; i<l;i++) {
      x++;
  }
  ```
 
  ```js
  for(var i = 0;i<l; i++) {
      x++;
  }
  ```

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/7700285-rule-require-spaces-in-between-for-statement?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gero3**
Fixes #912

Requires spaces in between for statement.

Type: Boolean

Values: true to requires spaces in between for statement.

#### Example

  "requireSpacesInForStatement": true
##### Valid

  for(var i = 0; i<l; i++) {
      x++;
  }
##### Invalid

  for(var i = 0;i<l;i++) {
      x++;
  }
  for(var i = 0; i<l;i++) {
      x++;
  }
  for(var i = 0;i<l; i++) {
      x++;
  }

### CODE REVIEW

**Comment by markelog** [comment]
Could add an opposite rule? 

And add this rule to the following presets:  "jquery", "wikimedia", "google", "airbnb" and "crockford".
Each preset changes with different commit.

File: lib/rules/require-spaces-in-for-statement.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        );
+    },
+
+    getOptionName: function() {
+        return 'requireSpacesInForStatement';
+    },
+
+    check: function(file, errors) {
+        file.iterateNodesByType('ForStatement', function(node) {
+            if (node.test) {
```
Comment (markelog) [inline lib/rules/require-spaces-in-for-statement.js]: I guess you can use assertion API here, should much cleaner
module called "<a href="https://github.com/jscs-dev/node-jscs/blob/8bd2ffd7bd1fc53a49d017f06322b178124534fa/lib/token-assert.js">token-assert</a>",
use example - https://github.com/jscs-dev/node-jscs/blob/8bd2ffd7bd1fc53a49d017f06322b178124534fa/lib/rules/validate-parameter-separator.js#L67-72

File: test/rules/require-spaces-in-for-statement.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            checker.configure({ requireSpacesInForStatement: true });
+        });
+        it.skip('should report missing spaces in for statement', function() {
+            assert(checker.checkString('for(i=0;i<l;i++){}').getErrorCount() === 2);
+        });
+        it('should report missing spaces in for statement before test statement', function() {
+            assert(checker.checkString('for(i=0; i<l;i++){}').getErrorCount() === 1);
+        });
+        it('should report missing spaces in for statement behind test statement', function() {
+            assert(checker.checkString('for(i=0;i<l; i++){}').getErrorCount() === 1);
```
Comment (markelog) [inline test/rules/require-spaces-in-for-statement.js]: Would be cool to check for case without initialization part, like:
```js
for( ; i<l; i++ ){}
```

File: test/rules/require-spaces-in-for-statement.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    beforeEach(function() {
+        checker = new Checker();
+        checker.registerDefaultRules();
+    });
+
+    describe('true', function() {
+        beforeEach(function() {
+            checker.configure({ requireSpacesInForStatement: true });
+        });
+        it.skip('should report missing spaces in for statement', function() {
```
Comment (markelog) [inline test/rules/require-spaces-in-for-statement.js]: Why?

File: lib/rules/require-spaces-in-for-statement.js (hunk 1 above)
Comment (gero3) [inline lib/rules/require-spaces-in-for-statement.js]: updated

File: test/rules/require-spaces-in-for-statement.js (hunk 2 above)
Comment (gero3) [inline test/rules/require-spaces-in-for-statement.js]: no problem

File: test/rules/require-spaces-in-for-statement.js (hunk 3 above)
Comment (gero3) [inline test/rules/require-spaces-in-for-statement.js]: now with extra info

**Comment by gero3** [comment]
Next commit will add the disallow rule.

I also added the rule to the specified presets and to MDCS.

**Comment by markelog** [comment]
> and to MDCS

Didn't find the mentions about it in the mdcs code-style, could you point me to it?

File: presets/wikimedia.json [hunk 4]
```diff
@@ -65,5 +65,6 @@
     "disallowNewlineBeforeBlockStatements": true,
     "validateLineBreaks": "LF",
     "validateQuoteMarks": "'",
-    "validateIndentation": "\t"
+    "validateIndentation": "\t",
+    "requireSpacesInForStatement": true
```
Comment (markelog) [inline presets/wikimedia.json]: /cc @Krinkle

File: test/rules/require-spaces-in-for-statement.js (hunk 3 above)
Comment (markelog) [inline test/rules/require-spaces-in-for-statement.js]: Nuh, i meant why do you `skip` that test?

File: lib/rules/require-spaces-in-for-statement.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                errors.assert.whitespaceBetween({
+                    token: prevToken1,
+                    nextToken: testToken,
+                    spaces: 1,
+                    message: 'Missing space after semicolon'
+                });
+            }
+            if (node.update) {
+                var updateToken = file.getTokenByRangeStart(node.update.range[0]);
+                var prevToken2 = file.getPrevToken(updateToken);
```
Comment (markelog) [inline lib/rules/require-spaces-in-for-statement.js]: `prevToken2 ` - not a fun of this kind of naming, maybe something more meaningful would be better?

**Comment by gero3** [comment]
https://github.com/mrdoob/three.js/wiki/Mr.doob%27s-Code-Style%E2%84%A2#for 

Altough it is not very clear it does mention the space behind the semicolon.

File: test/rules/require-spaces-in-for-statement.js (hunk 3 above)
Comment (gero3) [inline test/rules/require-spaces-in-for-statement.js]: copy past error

File: lib/rules/require-spaces-in-for-statement.js (hunk 5 above)
Comment (gero3) [inline lib/rules/require-spaces-in-for-statement.js]: fixed this

**Comment by gero3** [comment]
Also added the disallow.

File: presets/wikimedia.json (hunk 4 above)
Comment (Krinkle) [inline presets/wikimedia.json]: This seems already covered by `requireSpacesInsideParentheses`. Without that rule, we'd need a rule for `if`, `while`, `switch`, etc. as well. Not just `for`.

Comment (markelog) [inline presets/wikimedia.json]: Look at example a bit more:

```js
// this is valid:
for( var i = 0; i<l; i++ )

// this is isn't:
for( var i = 0;i<l;i++ )
```

**Comment by markelog** [comment]
@gero3 could you put preset changes in the different commit? That might be helpful in the future.

File: presets/wikimedia.json (hunk 4 above)
Comment (Krinkle) [inline presets/wikimedia.json]: @markelog Ah, I see. Nice! I still feel like this should maybe be generalised in some way, but sure, feel free to add this to the wikimedia preset.

**Comment by mikesherov** [comment]
LGTM

**Pull request closed without merging**